sanic/errorpages.py

```diff
@@ -435,94 +434,67 @@ def exception_response(
-
-            # Lastly, if there is an Accept header, make sure
-            # our choice is okay
-            if acceptable:
-                type_ = CONTENT_TYPE_BY_RENDERERS.get(renderer)  # type: ignore
-                if type_ and not acceptable.match(type_):
-                    # If the renderer selected is not in the Accept header
-                    # look through what is in the Accept header, and select
-                    # the first option that matches. Otherwise, just drop back
-                    # to the original default
-                    for accept in acceptable:
-                        mtype = f"{accept.type}/{accept.subtype}"
-                        maybe = RENDERERS_BY_CONTENT_TYPE.get(mtype)
-                        if maybe:
-                            renderer = maybe
-                            break
-                    else:
-                        renderer = base
+        mt = guess_mime(request, fallback)
+        renderer = RENDERERS_BY_CONTENT_TYPE.get(mt, base)
 
     renderer = t.cast(t.Type[BaseRenderer], renderer)
     return renderer(request, exception, debug).render()
+
+
+def guess_mime(req: Request, fallback: str) -> str:
+    # Attempt to find a suitable MIME format for the response.
+    # Insertion-ordered map of formats["html"] = "source of that suggestion"
+    formats = {}
+    name = ""
+    # Route error_format (by magic from handler code if auto, the default)
+    if req.route:
+        name = req.route.name
+        f = req.route.extra.error_format
+        if f in MIME_BY_CONFIG:
+            formats[f] = name
+
+    if not formats and fallback in MIME_BY_CONFIG:
+        formats[fallback] = "FALLBACK_ERROR_FORMAT"
+
+    # If still not known, check for the request for clues of JSON
+    if not formats and req.accept.match(JSON):
+        if JSON in req.accept:  # Literally, not wildcard
+            formats["json"] = "request.accept"
+        elif JSON in req.headers.getone("content-type", ""):
+            formats["json"] = "content-type"
+        # DEPRECATION: Remove this block in 24.3
+        else:
+            c = None
+            try:
+                c = req.json
+            except BadRequest:
+                pass
+            if c:
+                formats["json"] = "request.json"
+                deprecation(
+                    "Response type was determined by the JSON content of "
+                    "the request. This behavior is deprecated and will be "
+                    "removed in v24.3. Please specify the format either by\n"
+                    f'  error_format="json" on route {name}, by\n'
+                    '  FALLBACK_ERROR_FORMAT = "json", or by adding header\n'
+                    "  accept: application/json to your requests.",
+                    24.3,
+                )
+
+    # Any other supported formats
+    for k in MIME_BY_CONFIG:
+        if k not in formats:
+            formats[k] = "any"
```

Comment:
I feel like we should skip this test if the format was explicitly set:

```python
@app.get("/", error_format="json")
async def handler(request: Request):
    1 / 0
```

This PR does a much better job giving `FALLBACK_ERROR_FORMAT` its intended impact: the format to use when we are not sure what else to use. But, one thing that I feel we are still not covering is when the developer explicitly declares the format.

In the above scenario, we should only ever use JSON. Taking this step then begs the next: an API to explicitly declare a definitive format to use for everything. Right now to do this requires a mess of `@app.exception` or building a custom `ErrorHandler`.

Happy to take that up in a subsequent PR.

Reply:
Agreed, leave out "any" if explicitly set on route.

Reply:
I suppose this is to force JSON response to browser viewing a page on such route? Despite my previous comment, I am not entirely convinced of that being preferable even when manually defined on the route. I'll leave that up to you, but I prefer simplicity in documentation and code if adding additional logic doesn't serve a clear purpose.

Reply:
Fine with that for now. But let's do that skip for explicitly set, and then I think this is ready.

Reply:
Used fallback format for that instead. If set to auto, client preference is respected (via any format) as before, but apps can set that manually to get JSON responses while viewing in browser. Seems like the right way to handle this.